### PR TITLE
Week 9

### PR DESCRIPTION
**Pull request by Dev1LtJ**
Changes:
- refresh-page fetching usersList and tasksList from firebase was hoisted to "Main" Component;
- refresh-page fetching was rewritten using React & React-Redux Hooks;
- two packages of custom Hooks was created using React Hooks for training;
- Modals: CreateTrack, EditTrack, CreateTask, EditTask, CreateUser, EditUser were rewritten using hand-crafted Hooks (CreateTrack, EditTrack, CreateUser, EditUser with first package of custom Hooks, as well as CreateTask, EditTask with second package of custom Hooks);
- unnecessary now helpers were deleted.

### CODE REVIEW

File: src/components/Modals/Tasks/EditTask.jsx [hunk 1]
```diff
@@ -1,159 +1,114 @@
...
-          targetType,
-        }),
-      debounce(() => {
-        this.validateFields(name, value, targetType);
-      }, 1000),
-    );
-  }
+  useEffect(() => {
+    if (firstRender.current) {
+      firstRender.current = false;
```
Comment (maximsan) [inline src/components/Modals/Tasks/EditTask.jsx:34]: why do you need it ? it looks like your implementation is wrong if you need this in your case

File: src/components/Modals/Tasks/EditTask.jsx [hunk 2]
```diff
@@ -1,159 +1,114 @@
...
 
-  validateForm() {
-    const state = reducerFunc(this.state, {
-      type: EDIT_TASK_VALIDATE_FORM,
-    });
-    this.setState(state);
-  }
+  useEffect(() => {
+    dispatchValidator({ type: TASK_VALIDATE, payload: { state, event: onChangeSnapshot.current } });
+  }, [state.title, state.startDate, state.deadLine, state.selectedUsers]);
```
Comment (maximsan) [inline src/components/Modals/Tasks/EditTask.jsx:43]: Dependencies array should contains only used parameters. That is wrong

File: src/components/Modals/Tasks/EditTask.jsx [hunk 3]
```diff
@@ -1,159 +1,114 @@
...
+            onChange={handleChange}
+            error={validator.errors.selectedUsersError}
+          />
+        </div>
+        <div className={classes.requiredwarning}>* - these fields are required.</div>
+        <div className={classes.buttons}>
+          <Button onClick={editTask} roleClass='edit' disabled={!isValid}>
+            Edit
+          </Button>
+          <Button onClick={() => closeFunc()}>Close</Button>
```
Comment (maximsan) [inline src/components/Modals/Tasks/EditTask.jsx:107]: create a method

File: src/components/Modals/Tracks/CreateTrack.jsx [hunk 4]
```diff
@@ -1,136 +1,70 @@
...
-    };
-    this.liftUpCreateTrack = this.liftUpCreateTrack.bind(this);
-    this.closeModal = this.closeModal.bind(this);
-    this.onChange = this.onChange.bind(this);
-    this.validateFields = this.validateFields.bind(this);
-    this.validateForm = this.validateForm.bind(this);
-  }
+export default function CreateTrack({ task: { title, startDate }, closeFunc, liftUpCreateTrack }) {
+  const { state, onChange } = useInput({
+    id: createElemRefOnDB(TRACKS).id,
```
Comment (maximsan) [inline src/components/Modals/Tracks/CreateTrack.jsx:11]: looks like your method can be called outside of the component and you will avoid extra work

File: src/components/Modals/Tracks/CreateTrack.jsx [hunk 5]
```diff
@@ -1,136 +1,70 @@
...
-      task: { startDate, title },
-    } = this.props;
-    this.setState({ title, startDate });
-  }
+  const { errors, isValid } = useValidator(
+    {
+      name: false,
+      date: false,
+    },
+    undefined,
```
Comment (maximsan) [inline src/components/Modals/Tracks/CreateTrack.jsx:22]: it's a sign that your method API is bad

File: src/components/Modals/Tracks/CreateTrack.jsx [hunk 6]
```diff
@@ -1,136 +1,70 @@
...
+            onChange={onChange}
+          />
+          <CraftInput title='Track Note' id='note' type='textarea' value={state.note} onChange={onChange} />
+        </div>
+        <div className={classes.requiredwarning}>* - these fields are required.</div>
+        <div className={classes.buttons}>
+          <Button onClick={createTrack} roleClass='create' disabled={!isValid}>
+            Create
+          </Button>
+          <Button onClick={() => closeFunc()}>Close</Button>
```
Comment (maximsan) [inline src/components/Modals/Tracks/CreateTrack.jsx:63]: create a method

File: src/components/Modals/modals-helpers.js [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    default:
+      return prevState;
+  }
+};
+
+export const validatorReducer = (prevState, action) => {
+  switch (action.type) {
+    case TASK_VALIDATE:
+      if (action.payload.event.type === 'checkbox') {
+        const validity = !!action.payload.state.selectedUsers.length;
```
Comment (maximsan) [inline src/components/Modals/modals-helpers.js:111]: what does the name mean ? it's bad name from my point of view, you should rename it

File: src/components/Modals/modals-helpers.js [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+            [name]: validity,
+          },
+          errors: {
+            ...prevState.errors,
+            [`${name}Error`]: errorMsg,
+          },
+        };
+      }
+
+      return prevState;
```
Comment (maximsan) [inline src/components/Modals/modals-helpers.js:147]: why do you need this return ? I guess it is redundant

File: src/components/Modals/modals-helpers.js [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+            ...prevState.validator,
+            selectedUsers: validity,
+          },
+          errors: {
+            ...prevState.errors,
+            selectedUsersError: errorMsg,
+          },
+        };
+      }
+      if (action.payload.event.name in prevState.validator) {
```
Comment (maximsan) [inline src/components/Modals/modals-helpers.js:126]: hm, I don't like this approach. can we rework it ?

File: src/components/Modals/modals-helpers.js [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+};
+
+export const TASK_ONCHANGE = 'TASK_ONCHANGE';
+export const TASK_VALIDATE = 'TASK_VALIDATE';
+
+export const stateReducer = (prevState, action) => {
+  switch (action.type) {
+    case TASK_ONCHANGE:
+      if (action.payload.type === 'checkbox') {
+        if (prevState.selectedUsers.find((id) => id === action.payload.name)) {
```
Comment (maximsan) [inline src/components/Modals/modals-helpers.js:85]: save result in a variable before this

File: src/components/Modals/modals-helpers.js [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+  };
+};
+
+export const TASK_ONCHANGE = 'TASK_ONCHANGE';
+export const TASK_VALIDATE = 'TASK_VALIDATE';
+
+export const stateReducer = (prevState, action) => {
+  switch (action.type) {
+    case TASK_ONCHANGE:
+      if (action.payload.type === 'checkbox') {
```
Comment (maximsan) [inline src/components/Modals/modals-helpers.js:84]: I'd suggest to do payload destructiring here

File: src/components/Modals/modals-helpers.js [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+  const [state, setState] = useState(() => initialState);
+
+  const onChange = ({ target: { name, value } }) =>
+    setState((prevState) => {
+      return {
+        ...prevState,
+        [name]: value,
+      };
+    });
+  console.log(state);
```
Comment (maximsan) [inline src/components/Modals/modals-helpers.js:15]: remove

File: src/components/Modals/modals-helpers.js [hunk 13]
```diff
@@ -0,0 +1,151 @@
+import { useState, useEffect, useRef } from 'react';
+import { checkAllFormValidity, validateInput } from '../../utilities/form-validators';
+
+export const useInput = (initialState) => {
+  // why lazy loading doesn't work or why it's recreates every render?
+  const [state, setState] = useState(() => initialState);
+
+  const onChange = ({ target: { name, value } }) =>
```
Comment (maximsan) [inline src/components/Modals/modals-helpers.js:8]: wrap in useCallback

File: src/components/Modals/modals-helpers.js [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+      setErrors((prevErrors) => ({
+        ...prevErrors,
+        [`${name}Error`]: errorMsg,
+      }));
+
+      prevTrackedFieldsRef.current = trackedFields;
+    }
+    // Previous validator & errors (actual will be after useEffect's called render).
+    console.log(validator);
+    console.log(errors);
```
Comment (maximsan) [inline src/components/Modals/modals-helpers.js:57]: remove

File: src/components/Modals/modals-helpers.js [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+    });
+  console.log(state);
+
+  return {
+    state,
+    onChange,
+  };
+};
+
+export const useValidator = (initialState, password = '', startDate = '', ...trackedFields) => {
```
Comment (maximsan) [inline src/components/Modals/modals-helpers.js:23]: use an object for parameters in this case

File: src/components/Modals/modals-helpers.js [hunk 16]
```diff
@@ -0,0 +1,151 @@
+import { useState, useEffect, useRef } from 'react';
+import { checkAllFormValidity, validateInput } from '../../utilities/form-validators';
+
+export const useInput = (initialState) => {
+  // why lazy loading doesn't work or why it's recreates every render?
+  const [state, setState] = useState(() => initialState);
```
Comment (maximsan) [inline src/components/Modals/modals-helpers.js:6]: I'd suggest you to move lazy loading in place where you call custom hook and do check if it's function or an object and then call or not

File: src/components/Modals/modals-helpers.js [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+export const useValidator = (initialState, password = '', startDate = '', ...trackedFields) => {
+  const [validator, setValidator] = useState(() => initialState);
+  const [errors, setErrors] = useState(() => {
+    return Object.keys(initialState).reduce((result, key) => {
+      const errorKey = `${key}Error`;
+
+      return { ...result, [errorKey]: '' };
+    }, {});
+  });
+  const [isValid, setIsValid] = useState(() => false);
```
Comment (maximsan) [inline src/components/Modals/modals-helpers.js:32]: it doesn't make sense, set false

File: src/components/Modals/modals-helpers.js [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+  const prevTrackedFieldsRef = useRef(trackedFields);
+  const firstRender = useRef(true);
+
+  useEffect(() => {
+    const changedField = trackedFields.find((field, index) => field[1] !== prevTrackedFieldsRef.current[index][1]);
```
Comment (maximsan) [inline src/components/Modals/modals-helpers.js:37]: don't use magic numbers, create a helper for this and cover with unit tests

**Review by maximsan — CHANGES_REQUESTED** [review]
Hello! I left some comments. Let me know if you have questions. Good work!

File: src/components/Main/Main.jsx [hunk 19]
```diff
@@ -1,9 +1,30 @@
...
+
+  const dispatch = useDispatch();
+
+  useEffect(() => {
+    dispatch(fetchUsers());
+  }, [dispatch]);
+
+  useEffect(() => {
+    dispatch(fetchTasks());
+  }, [dispatch]);
```
[the remark this answers is not in the document]
Comment (Dev1LtJ) [reply src/components/Main/Main.jsx:22]: Fixed:
```suggestion
  useEffect(() => {
    dispatch(fetchUsers());
    dispatch(fetchTasks());
  }, [dispatch]);
```

File: src/components/Modals/Tasks/CreateTask.jsx [hunk 20]
```diff
@@ -1,168 +1,114 @@
...
-    };
-    this.onChange = this.onChange.bind(this);
-    this.liftUpCreateTask = this.liftUpCreateTask.bind(this);
-    this.closeModal = this.closeModal.bind(this);
-    this.validateFields = this.validateFields.bind(this);
-    this.validateForm = this.validateForm.bind(this);
-  }
+export default function CreateTask({ closeFunc, usersList, liftUpCreateTask }) {
+  const newTaskRef = useRef(createElemRefOnDB(TASKS));
+  const onChangeSnapshot = useRef({});
```
[the remark this answers is not in the document]
Comment (Dev1LtJ) [reply src/components/Modals/Tasks/CreateTask.jsx:12]: Fixed:
```suggestion
  const onChangeSnapshotRef = useRef({});
```

File: src/components/Modals/Tasks/CreateTask.jsx [hunk 21]
```diff
@@ -1,168 +1,114 @@
...
-    this.setState((prevState) => ({
-      ...prevState,
-      data: {
-        ...prevState.data,
-        id: prevState.newTaskRef.id,
-      },
-      usersList,
-    }));
-  }
+  const [state, dispatchState] = useReducer(stateReducer, newTaskRef.current.id, (newTaskId) => ({
```
[the remark this answers is not in the document]
Comment (Dev1LtJ) [reply src/components/Modals/Tasks/CreateTask.jsx:14]: According to useReducer specification, newTaskId (parameter of initializer) comes from newTaskRef.current.id (initialize value). newTaskRef.current is the reference (!), created in the firebase by the method: createElemRefOnDB(TASKS):
`export const createElemRefOnDB = (collection) => db.collection(collection).doc();`

File: src/components/Modals/Tasks/EditTask.jsx (hunk 3 above)
Comment (Dev1LtJ) [reply src/components/Modals/Tasks/EditTask.jsx:107]: Fixed:
```suggestion
          <Button onClick={closeModal}>Close</Button>
```

File: src/components/Modals/Tracks/CreateTrack.jsx (hunk 6 above)
Comment (Dev1LtJ) [reply src/components/Modals/Tracks/CreateTrack.jsx:63]:
```suggestion
          <Button onClick={closeFunc}>Close</Button>
```

File: src/components/Modals/modals-helpers.js (hunk 8 above)
Comment (Dev1LtJ) [reply src/components/Modals/modals-helpers.js:147]: As we have discussed this case on the meeting, arrow function requires return.

File: src/components/Modals/modals-helpers.js (hunk 12 above)
Comment (Dev1LtJ) [reply src/components/Modals/modals-helpers.js:15]: Removed:
```suggestion

```

File: src/components/Modals/modals-helpers.js (hunk 14 above)
Comment (Dev1LtJ) [reply src/components/Modals/modals-helpers.js:57]: Removed:
```suggestion

```

File: src/components/Modals/modals-helpers.js (hunk 17 above)
Comment (Dev1LtJ) [reply src/components/Modals/modals-helpers.js:32]: fixed:
```suggestion
  const [isValid, setIsValid] = useState(false);
```

File: src/components/Modals/modals-helpers.js (hunk 16 above)
Comment (Dev1LtJ) [reply src/components/Modals/modals-helpers.js:6]: Fixed:
```suggestion
const [state, setState] = useState(typeof initialState === 'function' ? initialState() : initialState);
```

File: src/components/Modals/modals-helpers.js (hunk 13 above)
Comment (Dev1LtJ) [reply src/components/Modals/modals-helpers.js:8]: Wrapped:

```suggestion
  const onChange = useCallback(
    ({ target: { name, value } }) =>
      setState((prevState) => {
        return {
          ...prevState,
          [name]: value,
        };
      }),
    [],
  );
```

File: src/components/Modals/Tracks/CreateTrack.jsx (hunk 4 above)
Comment (Dev1LtJ) [reply src/components/Modals/Tracks/CreateTrack.jsx:11]: I created a ref: ` const newTrackRef = useRef(createElemRefOnDB(TRACKS));`
```suggestion
  const { state, onChange } = useInput(() => ({
    id: newTrackRef.current.id,
    name: '',
    date: '',
    note: '',
  }));
```

File: src/components/Modals/modals-helpers.js (hunk 11 above)
Comment (Dev1LtJ) [reply src/components/Modals/modals-helpers.js:84]: Rewritten:
```suggestion
export const stateReducer = (prevState, action) => {
  const {
    payload: { name, value, type },
  } = action;

  switch (action.type) {
    case TASK_ONCHANGE:
      if (type === 'checkbox') {
        if (prevState.selectedUsers.find((id) => id === name)) {
          return {
            ...prevState,
            selectedUsers: prevState.selectedUsers.filter((id) => id !== name),
          };
        }
```

File: src/components/Modals/modals-helpers.js (hunk 15 above)
Comment (Dev1LtJ) [reply src/components/Modals/modals-helpers.js:23]: Reworked:
```suggestion
export const useValidator = (initialState, password = '', startDate = '', trackedFields) => {
```

File: src/components/Modals/modals-helpers.js (hunk 18 above)
Comment (Dev1LtJ) [reply src/components/Modals/modals-helpers.js:37]: Created:
```suggestion
    const changedField = findChangedField(prevTrackedFieldsRef.current, trackedFields);
```

File: src/components/Modals/modals-helpers.js (hunk 7 above)
Comment (Dev1LtJ) [reply src/components/Modals/modals-helpers.js:111]: Fixed:
```suggestion
        const isValid = !!action.payload.state.selectedUsers.length;
        const errorMsg = isValid ? '' : 'At least one user must be selected';
```

File: src/components/Modals/modals-helpers.js (hunk 10 above)
Comment (Dev1LtJ) [reply src/components/Modals/modals-helpers.js:85]: Saved:
```suggestion
        const isSelectedUserExists = prevState.selectedUsers.find((id) => id === name);
        if (isSelectedUserExists) {
```

File: src/components/Modals/modals-helpers.js (hunk 9 above)
Comment (Dev1LtJ) [reply src/components/Modals/modals-helpers.js:126]: Reworked: created method for search: `const isFieldNameInValidator = Object.keys(prevState.validator).includes(name);`
So:
```suggestion
      if (isFieldNameInValidator) {
```

File: src/components/Modals/Tasks/EditTask.jsx (hunk 1 above)
Comment (Dev1LtJ) [reply src/components/Modals/Tasks/EditTask.jsx:34]: Reworked with creating custom hook:
```suggestion
export const useAllSelectedFormsValidityChecker = (validator, initialState) => {
  const [isValid, setIsValid] = useState(false);
  const initialStateRef = useRef(initialState);

  const isOneOfStateFieldsChanged = useMemo(() => checkIfOneOfFieldsChanged(initialStateRef.current, initialState), [
    initialStateRef.current,
    initialState,
  ]);
  useEffect(() => {
    if (isOneOfStateFieldsChanged) {
      console.log('useEffect setIsValid');
      setIsValid(checkAllFormValidity(validator));
    }
  }, [validator, isOneOfStateFieldsChanged]);

  return isValid;
};
```

File: src/components/Modals/Tasks/EditTask.jsx (hunk 2 above)
Comment (Dev1LtJ) [reply src/components/Modals/Tasks/EditTask.jsx:43]: Fixed:
```suggestion
  useEffect(() => {
    dispatchValidator({ type: TASK_VALIDATE, payload: { state, event: onChangeSnapshotRef.current } });
    /*
     I don't need other dependencies here, coz validator's reducer has a checker, which verifies,
      event-changed field inside or outside validator and returns prevState.
    */
  }, [state]);
```

File: src/components/Modals/Tracks/CreateTrack.jsx (hunk 5 above)
Comment (Dev1LtJ) [reply src/components/Modals/Tracks/CreateTrack.jsx]: Reworked

File: src/components/Modals/modals-helpers.js [hunk 22]
```diff
@@ -0,0 +1,179 @@
...
+    state,
+    onChange,
+  };
+};
+
+export function carriedUseValidator(initialState) {
+  return (password = '') => {
+    return (startDate = '') => {
+      return (trackedFields) => {
+        return useValidator(initialState, password, startDate, trackedFields);
```
Comment (maximsan) [inline src/components/Modals/modals-helpers.js:30]: now it looks scary. just use wrapper with default values for unnecessary fields

File: src/utilities/form-helpers.js [hunk 23]
```diff
@@ -1,2 +1,16 @@
 export const getLowerCasedStr = (str) => str.toLowerCase();
 export const getTrimmedStr = (str) => str.trim();
+
+export const findChangedField = (prevObj, newObj) => {
+  const prevObjEntries = Object.entries(prevObj);
+  const newObjEntries = Object.entries(newObj);
+
+  return newObjEntries.find((field, index) => field[1] !== prevObjEntries[index][1]);
```
Comment (maximsan) [inline src/utilities/form-helpers.js:8]: add explanation about why we use this approach and why we use [1] in first and second object

File: src/components/Modals/Tracks/CreateTrack.jsx [hunk 24]
```diff
@@ -1,136 +1,70 @@
...
+  }));
 
-  validateForm() {
-    const state = reducerFunc(this.state, { type: CREATE_TRACK_VALIDATE_FORM });
-    this.setState(state);
-  }
+  const { errors, validator } = carriedUseValidator(() => ({
+    name: false,
+    date: false,
+  }))()(startDate)(useMemo(() => ({ name: state.name, date: state.date }), [state.name, state.date]));
```
Comment (maximsan) [inline src/components/Modals/Tracks/CreateTrack.jsx:22]: @Dev1LtJ  I don't see any real reason to use `useMemo` here and in other cases, just pass values. That should be fine. But actually I left in this hook that it's better to rewrite it. it looks scary :) it should be covered by unit-tests ^-^

**Review by maximsan — CHANGES_REQUESTED** [review]
Hello! Left some comments. Please review, let me know fi you have questions. Thanks! Good try :) It's almost better than good )

File: src/components/Modals/modals-helpers.js [hunk 25]
```diff
@@ -0,0 +1,179 @@
...
+    validator,
+  };
+};
+
+export const useAllSelectedFormsValidityChecker = (validator, initialState) => {
+  const [isValid, setIsValid] = useState(false);
+  const initialStateRef = useRef(initialState);
+
+  const isOneOfStateFieldsChanged = useMemo(() => checkIfOneOfFieldsChanged(initialStateRef.current, initialState), [
+    initialStateRef.current,
```
[the remark this answers is not in the document]
Comment (Dev1LtJ) [reply src/components/Modals/modals-helpers.js:79]: Removed
```suggestion

```

File: src/utilities/form-helpers.js (hunk 23 above)
Comment (Dev1LtJ) [reply src/utilities/form-helpers.js:8]: Added:
```suggestion
export const findChangedField = (prevObj, newObj) => {
  const prevObjEntries = Object.entries(prevObj);
  const newObjEntries = Object.entries(newObj);
  // it's a method to safe comparing two objects. [1] - it is the value.
  return newObjEntries.find((field, index) => field[1] !== prevObjEntries[index][1]);
};
```

File: src/components/Modals/Tasks/CreateTask.jsx [hunk 26]
```diff
@@ -1,168 +1,121 @@
...
+            isRequired
+            id='selectedUsers'
+            type='checkbox'
+            value={state.selectedUsers}
+            options={usersList}
+            onChange={handleChange}
+            error={validator.errors.selectedUsersError}
+          />
+        </div>
+        <div className={classes.requiredwarning}>* - these fields are required.</div>
```
[the remark this answers is not in the document]
Comment (Dev1LtJ) [reply src/components/Modals/Tasks/CreateTask.jsx:109]: Fixed: 
```suggestion
        <div className={classes.requiredWarning}>* - these fields are required.</div>
```

File: src/components/Modals/modals-helpers.js [hunk 27]
```diff
@@ -0,0 +1,179 @@
...
+
+export const useValidator = (initialState, password, startDate, trackedFields) => {
+  const [validator, setValidator] = useState(typeof initialState === 'function' ? initialState() : initialState);
+  const [errors, setErrors] = useState(() => {
+    return Object.keys(validator).reduce((result, key) => {
+      const errorKey = `${key}Error`;
+
+      return { ...result, [errorKey]: '' };
+    }, {});
+  });
```
[the remark this answers is not in the document]
Comment (Dev1LtJ) [reply src/components/Modals/modals-helpers.js:44]: Created helper:
```suggestion
  const [errors, setErrors] = useState(() => createErrorsStateByValidator(validator));
```
Also covered by tests.

File: src/components/Modals/Tasks/CreateTask.jsx [hunk 28]
```diff
@@ -1,168 +1,121 @@
...
-          name,
-          value,
-          targetType,
-        }),
-      debounce(() => {
-        this.validateFields(name, value, targetType);
-      }, 1000),
-    );
-  }
+  const [validator, dispatchValidator] = useReducer(validatorReducer, undefined, () => ({
```
[the remark this answers is not in the document]
Comment (Dev1LtJ) [reply src/components/Modals/Tasks/CreateTask.jsx:28]:
```suggestion
  const [{ validator, errors }, dispatchValidator] = useReducer(validatorReducer, undefined, () => ({
```

File: src/components/Modals/modals-helpers.js (hunk 22 above)
Comment (Dev1LtJ) [reply src/components/Modals/modals-helpers.js:30]: Deleted.
```suggestion

```

File: src/components/Modals/Tracks/CreateTrack.jsx (hunk 24 above)
Comment (Dev1LtJ) [reply src/components/Modals/Tracks/CreateTrack.jsx:22]: Rewrited:
```suggestion
  const { errors, validator } = useValidator(
    () => ({
      name: false,
      date: false,
    }),
    undefined,
    startDate,
    { name: state.name, date: state.date },
  );
```

File: src/components/Modals/modals-helpers.js [hunk 29]
```diff
@@ -0,0 +1,179 @@
...
+  const [isValid, setIsValid] = useState(false);
+  const initialStateRef = useRef(initialState);
+
+  const isOneOfStateFieldsChanged = useMemo(() => checkIfOneOfFieldsChanged(initialStateRef.current, initialState), [
+    initialStateRef.current,
+    initialState,
+  ]);
+  useEffect(() => {
+    if (isOneOfStateFieldsChanged) {
+      console.log('useEffect setIsValid');
```
[the remark this answers is not in the document]
Comment (Dev1LtJ) [reply src/components/Modals/modals-helpers.js:84]: removed

File: src/components/Modals/modals-helpers.js [hunk 30]
```diff
@@ -0,0 +1,179 @@
...
+      return { ...result, [errorKey]: '' };
+    }, {});
+  });
+  const prevTrackedFieldsRef = useRef(trackedFields);
+
+  useEffect(() => {
+    const changedField = findChangedField(prevTrackedFieldsRef.current, trackedFields);
+
+    if (changedField) {
+      console.log('useEffect setValidator, setErrors');
```
[the remark this answers is not in the document]
Comment (Dev1LtJ) [reply src/components/Modals/modals-helpers.js:51]: removed

File: src/components/Modals/modals-helpers.js [hunk 31]
```diff
@@ -0,0 +1,179 @@
...
+    ({ target: { name, value } }) =>
+      setState((prevState) => {
+        return {
+          ...prevState,
+          [name]: value,
+        };
+      }),
+    [],
+  );
+  console.log('onChange');
```
[the remark this answers is not in the document]
Comment (Dev1LtJ) [reply src/components/Modals/modals-helpers.js:18]: removed

**Review by maximsan — APPROVED** [review]
Looks fine. Congratulations!

**Review by YauhenKoipish — APPROVED** [review]

**Pull request merged**